### PR TITLE
Implement a StateStore write-through cache.

### PR DESCRIPTION
**Pull request by nickbp**
Only as consistent as the underlying StateStore implementation.

### CODE REVIEW

File: src/main/java/org/apache/mesos/curator/CuratorStateStore.java [hunk 1]
```diff
@@ -386,7 +388,8 @@ public void clearProperty(final String key) throws StateStoreException {
         }
     }
 
-    void close() {
+    @VisibleForTesting
```
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/curator/CuratorStateStore.java]: Neat.

**Comment by gabrielhartmann** [comment]
@nickbp:  It looks like what is missing here is the ability to transactionally store both a TaskInfo and a TaskStatus simultaneously.  We really want this feature so we can guarantee that when a TaskInfo is present in the StateStore, that it also has a TaskStatus (initially TASK_STAGING).

**Comment by nickbp** [comment]
Yes this is just covering caching of the StateStore interface as-is and allow applications to avoid DIYing their own caching. In addition to this there are additional changes to be made in CuratorStateStore to make that more transactional/atomic, plus the earlier discussion about how to (re)initialize TaskStatus on TaskInfo changes and what form that should take

File: src/main/java/org/apache/mesos/state/StateStoreCache.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        } finally {
+            RLOCK.unlock();
+        }
+    }
+
+    @Override
+    public void storeTasks(Collection<TaskInfo> tasks) throws StateStoreException {
+        RWLOCK.lock();
+        try {
+            store.storeTasks(tasks);
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: Is it possible that storeTasks method persists tasks collection fully or partially and throws an exception ? If yes, then that particular case will make the cache inconsistent.

File: src/main/java/org/apache/mesos/state/StateStoreCache.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                        taskName, nameToId, idToTask.keySet()));
+            }
+            return Optional.of(task);
+        } finally {
+            RLOCK.unlock();
+        }
+    }
+
+    @Override
+    public Set<TaskStatus> getTaskStatuses() throws Exception {
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: Can we de-dup `getTaskStatuses` vs `fetchStatuses` ? Or there's a good reason why both should exist ?

File: src/main/java/org/apache/mesos/state/StateStoreCache.java (hunk 2 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: Yes. This will only be as consistent as the underlying state store.

Comment (mohitsoni) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: Okay.

File: src/main/java/org/apache/mesos/state/StateStoreCache.java (hunk 3 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: I agree, don't know why there need to be both. getTaskStatuses() is from TaskStatusProvider which the base StateStore interface extends.

We could either:
- Update TaskStatusProvider interface (Set<TaskInfo>) to align with StateStore interface (List<TaskInfo>)
- Update StateStore interface to align with TaskStatusProvider interface

File: src/main/java/org/apache/mesos/state/StateStoreCache.java (hunk 2 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: Filed INFINITY-570 for fixing CuratorStateStore

File: src/main/java/org/apache/mesos/state/StateStoreCache.java (hunk 3 above)
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: If there's no good reason for `TaskStatusProvider` to be around, maybe we can remove `TaskStatusProvider` in favor of just `StateStore` ? As `StateStore` provides the required functionality.

File: src/main/java/org/apache/mesos/state/StateStoreCache.java (hunk 2 above)
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: Thanks!

File: src/main/java/org/apache/mesos/state/StateStoreCache.java (hunk 3 above)
Comment (nickbp) [inline src/main/java/org/apache/mesos/state/StateStoreCache.java]: Done: Removed `TaskStatusProvider` interface and switched `DefaultReconciler` (its sole user) to use `StateStore` directly.

**Comment by mohitsoni** [comment]
Regarding @gabrielhartmann's comment on transactionally storing TaskInfo and TaskStatus:
@nickbp Are you going to update the CuratorStateStore to support that in this PR or you have planned additional work for it ?

**Comment by nickbp** [comment]
I was planning on just having a cache that works against the current interface, figuring that interface changes to StateStore itself would be done separately.

**Comment by nickbp** [comment]
Ping, any other comments?

**Pull request merged**